### PR TITLE
[DOCU-1401] Adds 2.4.0.0 beta change log entry

### PR DESCRIPTION
**Pull request by HCloward**
### Review
Combined both [CE 2.4.0.0 changelog](https://github.com/Kong/kong/blob/2.4.0-rc.1/CHANGELOG.md#240-rc1) and [EE - changelog](https://konghq.atlassian.net/wiki/spaces/FTT/pages/8356124/Kong+Enterprise+-+Changelog+Features+Release+Schedule#KongEnterprise-Changelog%2CFeatures%26ReleaseSchedule-KongEnterprise2.4.0.0Beta) for the 2.4.0 beta release.

[Google doc with original CE 2.4 beta changelog questions](https://docs.google.com/document/d/1wH08qMpYgzUWGyfn1Fyk1s9m_53sB423D1o79WqAQ0Q/edit)
### Summary
[DOCU-1401](https://konghq.atlassian.net/browse/DOCU-1401)
### Reason
Because we are releasing a 2.4.0.0 beta and want to inform users of the updates.
### Testing
[2.4.0.0-beta changelog](https://deploy-preview-2786--kongdocs.netlify.app/enterprise/changelog/#2400-beta)

### CODE REVIEW

File: app/enterprise/changelog.md [hunk 1]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  still connect to the control plane. Also, data planes are allowed to have a superset
+  of plugins in addition to the control plane plugins. [6932](https://github.com/Kong/kong/pull/6932)
+  <div class="alert alert-ee blue">
+    Data planes are not allowed to connect to control planes if they are a different major version,
+    a version newer than the control plane’s version, or missing plugins from the control plane.
+  </div>
+- This Kong version allows UTF-8 characters in tags. This update expands the range of
+  accepted characters in tags from a limited set of ASCII characters to almost all of UTF-8 sequences.
+  Exceptions:
+  - ',' and '/' are reserved for filtering tags with "and" and "or" and are not allowed in tags.
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:23]:
```suggestion
  - `,` and `/` are reserved for filtering tags with "and" and "or", and are not allowed in tags.
```

File: app/enterprise/changelog.md [hunk 2]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+#### Core
+- This Kong version includes relaxed version check in hybrid mode between control planes
+  and data planes, allowing data planes that are missing minor updates (up to two) to
+  still connect to the control plane. Also, data planes are allowed to have a superset
+  of plugins in addition to the control plane plugins. [6932](https://github.com/Kong/kong/pull/6932)
+  <div class="alert alert-ee blue">
+    Data planes are not allowed to connect to control planes if they are a different major version,
+    a version newer than the control plane’s version, or missing plugins from the control plane.
+  </div>
+- This Kong version allows UTF-8 characters in tags. This update expands the range of
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:20]:
```suggestion
- UTF-8 characters can now be used in tags. This update expands the range of
```

File: app/enterprise/changelog.md [hunk 3]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  accepted characters in tags from a limited set of ASCII characters to almost all of UTF-8 sequences.
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:21]:
```suggestion
  accepted characters in tags from a limited set of ASCII characters to almost all UTF-8 sequences.
```

File: app/enterprise/changelog.md [hunk 4]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  Exceptions:
+  - ',' and '/' are reserved for filtering tags with "and" and "or" and are not allowed in tags.
+  - Non-printable ASCII (like the space character) is not allowed.
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:24]:
```suggestion
  - Non-printable ASCII (for example, the space character) is not allowed.
```

File: app/enterprise/changelog.md [hunk 5]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+- This Kong version supports Online Certificate Status Protocol (OCSP) responder in cluster for
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:25]:
```suggestion
- Kong Gateway now supports using an Online Certificate Status Protocol (OCSP) responder in the cluster for
```

File: app/enterprise/changelog.md [hunk 6]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  hybrid mode control planes. This new feature can be configured in the `kong.conf` file.
+  [6887](https://github.com/Kong/kong/pull/6887)
+- In this Kong version, Postgres `ssl_version` configuration now supports 'any' and will negotiate with server.
+  This update ensures that `luasec` will negotiate the most secure protocol available if not otherwise set.
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:29]:
```suggestion
- Postgres `ssl_version` configuration now supports `any` and will negotiate with the server.
  If `ssl_version` is not set explicitly, this update ensures that `luasec` will negotiate the most secure protocol available.
```

Is it set to `any` by default now? I _think_ that's what this is saying. If true, then I'd rephrase to something like:

- Postgres `ssl_version` configuration now defaults to `any`. If `ssl_version` is not explicitly set, the `any` option ensures that `luasec` will negotiate the most secure protocol available.

File: app/enterprise/changelog.md [hunk 7]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+
+#### PDK
+- This release includes a new JavaScript Plugin Development Kit (PDK). This addition
+  allows users to write Kong plugins in JavaScript and TypeScript.
+- This release includes support for Protobuf plugin communication protocol, which can be used in
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:34]:
```suggestion
- This release includes support for the Protobuf plugin communication protocol, which can be used in
```

File: app/enterprise/changelog.md [hunk 8]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  place of MessagePack to communicate with non-Lua plugins. [6941](https://github.com/Kong/kong/pull/6941)
+- This release enables `ssl_certificate` phase on plugins in `stream` module.
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:36]:
```suggestion
- This release enables the `ssl_certificate` phase on plugins in the `stream` module.
```

File: app/enterprise/changelog.md [hunk 9]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  [6873](https://github.com/Kong/kong/pull/6873)
+
+#### Plugins
+- [OPA](/hub/kong-inc/opa) (`opa`)
+  - New OPA plugin forwards requests to an Open Policy Agent and processes the request
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:41]:
```suggestion
- **New plugin:** [OPA](/hub/kong-inc/opa) (`opa`)
  - The OPA plugin forwards requests to an Open Policy Agent and processes the request
```

File: app/enterprise/changelog.md [hunk 10]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+    only if the authorization policy allows for it. **Released as BETA and should not be deployed
+    in production environment.**
+- Mocking (`mocking`)
+  - New Kong Mocking plugin leverages standards-based Open API Specifications (OAS)
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:45]:
```suggestion
- **New plugin:** Mocking (`mocking`)
  - The Kong Mocking plugin leverages standards-based Open API Specifications (OAS)
```

File: app/enterprise/changelog.md [hunk 11]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+    the [Parameters](https://docs.konghq.com/hub/kong-inc/zipkin/#parameters) section of the Zipkin
+    plugin documentation for more information. [101](https://github.com/Kong/kong-plugin-zipkin/pull/101 )
+  - The plugin now allows insertion of custom tags on the Zipkin request trace. See `config.tags_header`
+    in the [Parameters](https://docs.konghq.com/hub/kong-inc/zipkin/#parameters) section of the Zipkin
+    plugin documentation for more information. [102](https://github.com/Kong/kong-plugin-zipkin/pull/102)
+  - The plugin now allows the creation of baggage items on child spans.
+    [98](https://github.com/Kong/kong-plugin-zipkin/pull/98)
+- [JWT](/hub/kong-inc/jwt) (`jwt`) 
+  - The plugin now supports ES384 JWTs signature validation. [6854](https://github.com/Kong/kong/pull/6854)
+- Several plugins including: [File Log](/hub/kong-inc/file-log), [Loggly](/hub/kong-inc/loggly),
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:59]:
```suggestion
- Logging plugins: [File Log](/hub/kong-inc/file-log), [Loggly](/hub/kong-inc/loggly),
```

File: app/enterprise/changelog.md [hunk 12]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  [Syslog](/hub/kong-inc/syslog), [TCP Log](/hub/kong-inc/tcp-log), [UDP Log](/hub/kong-inc/udp-log),
+  and [HTTP Log](/hub/kong-inc/http-log)
+  - Kong administrators now have a powerful new capability to transform logs to any format that’s needed
+    for capturing, indexing and correlating logs  When formatting, developers also have the ability to
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:63]:
```suggestion
    for capturing, indexing, and correlating logs. When formatting, developers also have the ability to
```

File: app/enterprise/changelog.md [hunk 13]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+    execute custom Lua code - opening up a new range of possibilities for generating logs to a specification
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:64]:
```suggestion
    execute custom Lua code, opening up a new range of possibilities for generating logs to whichever specification
```

File: app/enterprise/changelog.md [hunk 14]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+    that administrators need the most. This feature uses the new PDK method `kong.log.set_serialize_value`,
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:65]:
```suggestion
    administrators need the most. This feature uses the new PDK method `kong.log.set_serialize_value`,
```

File: app/enterprise/changelog.md [hunk 15]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  [6796](https://github.com/Kong/kong/pull/6796)
+- Schema validations now log more descriptive error messages when types are invalid.
+  [6593](https://github.com/Kong/kong/pull/6593)
+- Kong now ignores tags in Cassandra when filtering by multiple entities, which is the
+  expected behavior and the one already existent when using Postgres databases.
+  See [Tags](https://docs.konghq.com/enterprise/2.3.x/admin-api/#tags) in the Admin API
+  documentation for more information. [6931](https://github.com/Kong/kong/pull/6931)
+- Users can proxy websockets through Kong when using Chrome or Firefox. `HTTP Connection`
+  headers usually carry a list of headers meant for receivers (in this case a proxy) that
+  are not meant to be proxied further (hop-by-hop headers). The Kong gateway correctly
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:83]:
```suggestion
  are not meant to be proxied further (hop-by-hop headers). The Kong Gateway correctly
```

File: app/enterprise/changelog.md [hunk 16]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  cleans these headers, but it was a bit too aggressive and the`Upgrade` header was cleared
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:84]:
```suggestion
  cleans these headers, but it was a bit too aggressive and the `Upgrade` header was cleared
```

File: app/enterprise/changelog.md [hunk 17]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  from response `Connection` headers, causing the connections to fail when using Firefox
+  or Chrome. [6929](https://github.com/Kong/kong/pull/6929)
+- Kong now accepts fully-qualified domain names ending in dots to be compliant with RFC internet standards.
+  [6864](https://github.com/Kong/kong/pull/6864)
+- Kong has removed unnecessary load from DNS servers by resolving an issue occuring when
+  using upstreams for load balancing. When caching services, Kong does not warmup upstream
+  names used as service hosts entries when warming up DNS entries, as they are not real DNS
+  entries. [6891](https://github.com/Kong/kong/pull/6891)
+- Migrations order is now guaranteed to be always the same. Before the order in which the
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:93]:
```suggestion
- Migrations order is now guaranteed to be always the same. Before, the order in which the
```

File: app/enterprise/changelog.md [hunk 18]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  subsystems were loaded changed randomly. Now, subsystems are sorted alphabetically, so the
+  order in which migrations are executed is constant. [6901](https://github.com/Kong/kong/pull/6901)
+- Buffered responses are disabled on connection upgrades. Kong cannot do buffered responses with,
+  for example, websocket connection upgrades. [6902](https://github.com/Kong/kong/pull/6902)
+- Entity relationships are now traverse-order-independent. This fix keeps track of all copied
+  records and injects the new field on each copy.
+  [6843](https://github.com/Kong/kong/commit/5f2a87259e3b474ec6129490bba82dac0aeba1cf)
+- The host header is updated between balancer retries. Before Kong set the `Host` header during
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:101]:
```suggestion
- The host header is updated between balancer retries. Before, Kong Gateway set the `Host` header during
```

File: app/enterprise/changelog.md [hunk 19]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  to be retried. [6796](https://github.com/Kong/kong/pull/6796)
+- The router prioritizes the route with most matching headers when matching headers.
+  [6638](https://github.com/Kong/kong/pull/6638)
+- Fixed an issue involving multipart/form-data boundary checks. [6638](https://github.com/Kong/kong/pull/6638)
+- Fixed an issue that occurred when using upstreams for load balancing, where Kong was attempting
+  to resolve the upstream name instead of the hostname and failing with the errors `name resolution failed`
+  and `dns server error: 3 name error`. The following updates correct the issue:
+  - Kong does not cache empty upstream name dictionaries. [7002](https://github.com/Kong/kong/pull/7002)
+  - Kong does not assume upstreams don't exist after init phases. [7010](https://github.com/Kong/kong/pull/7010)
+- Opentracing libraries (opentracing, jaegar,datadog) bumped to latest versions.
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:112]:
```suggestion
- Opentracing libraries (opentracing, jaegar, datadog) bumped to latest versions.
```

File: app/enterprise/changelog.md [hunk 20]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+- The Developer Portal is now disabled when running without a license.
+
+#### PDK
+- Now Kong does not leave plugin servers alive after exiting and does not try to
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:116]:
```suggestion
- Kong Gateway no longer leaves plugin servers alive after exiting and does not try to
```
Not sure if this is the best phrasing, feel free to adjust.

File: app/enterprise/changelog.md [hunk 21]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  start them in the unsupported stream subsystem. [6849](https://github.com/Kong/kong/pull/6849)
+- Golang does not cache `kong.log` methods. The `kong` table has some special-case logic
+  for the `.log` subtable, and avoiding cacheing preserves this logic. [6701](https://github.com/Kong/kong/pull/6701)
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:119]:
```suggestion
  for the `.log` subtable, and avoiding caching preserves this logic. [6701](https://github.com/Kong/kong/pull/6701)
```

File: app/enterprise/changelog.md [hunk 22]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+- The `response` phase is now included on the list of public phases. [6638](https://github.com/Kong/kong/pull/6638)
+- Config file style and options case are now consistent. [6981](https://github.com/Kong/kong/pull/6981)
+- Added correct Protobuf MacOS path to enable external plugins in Homebrew installations. [6980](https://github.com/Kong/kong/pull/6980)
+- Now Kong auto-escapes upstream paths(`kong.service.request.set_path()`) to avoid proxying errors.
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:123]:
```suggestion
- Kong Gateway now auto-escapes upstream paths (`kong.service.request.set_path()`) to avoid proxying errors.
```

File: app/enterprise/changelog.md [hunk 23]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  [6978](https://github.com/Kong/kong/pull/6978)
+- In the Golang PDK, ports are now declared as `Int`. Before they were incorrectly declared
+  `Strings`. [6994](https://github.com/Kong/kong/pull/6994)
+
+#### Plugins
+- The [Rate Limiting](/hub/kong-inc/rate-limiting) and [Response Rate Limiting](/hub/kong-inc/response-ratelimiting)
+  plugins included default options that did not work well with DB-less or hybrid deployments mostly because the
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:130]:
```suggestion
  plugins included default options that did not work well with DB-less or hybrid deployments because the
```

File: app/enterprise/changelog.md [hunk 24]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  database was not available on gateway nodes. With this fix, the default values and validation rules are now
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:131]:
```suggestion
  database is not available on data plane nodes. With this fix, the default values and validation rules are now
```

File: app/enterprise/changelog.md [hunk 25]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  compatible with DB-less or hybrid modes. [6885](https://github.com/Kong/kong/pull/6885)
+- [OAuth 2.0](/hub/kong-inc/oauth2) (`oauth2`) 
+  - To improve user experience and limit confusion, the plugin now handles cases of client invalid
+    token generation in a more predictable way. Before the resulting error codes were confusing and
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:135]:
```suggestion
    token generation in a more predictable way. Before, the resulting error codes were confusing and
```

File: app/enterprise/changelog.md [hunk 26]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+    the access phase was skipped. [105](https://github.com/Kong/kong-plugin-zipkin/pull/105)
+- [OpenID Connect](/hub/kong-inc/openid-connect) (`openid-connect`)
+  - `kong-openid-connect` library updated to v2.1.0
+    - Updates `lua-resty-nettle` dependency to v2.0.
+    - Treats JWE tokens as opaque, so that they can be introspected.
+    - Adds support for Ed448 curve in EdDSA signing and verification and JWKS key generation.
+- [Kong JWT Signer](/hub/kong-inc/jwt-signer) (`jwt-signer`)
+  - Cache now uses upsert instead of insert/update with databases.
+  - Key rotation is now more resilient on errors.
+  - Adds Db-less improvements.
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:151]:
```suggestion
  - Adds DB-less improvements.
```

File: app/enterprise/changelog.md [hunk 27]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+- [Exit Transformer](/hub/kong-inc/exit-transformer) (`exit-transformer`)
+  - The plugin was not allowing access to Kong module within the sandbox, only to `kong.request`,
+    which prevented access to `kong.log` for example.
```
Comment (lena-larionova) [inline app/enterprise/changelog.md:154]:
```suggestion
  - The plugin was not allowing access to any Kong modules within the sandbox except `kong.request`,
    which prevented access to other necessary modules such as `kong.log`.
```

**Review by lena-larionova — APPROVED** [review]
LGTM from docs side.

File: app/enterprise/changelog.md [hunk 28]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+- [Zipkin](/hub/kong-inc/zipkin) (`zipkin`)
+  - The W3C parsing function was returning a non-used extra value that has been removed, and the plugin
+    now early-exits if there is a problem with the W3C Trace Context header.
+    [100](https://github.com/Kong/kong-plugin-zipkin/pull/100)
+  - Fixed a bug which randomly caused requests to fail with a 400 error code during span timestamping if
+    the access phase was skipped. [105](https://github.com/Kong/kong-plugin-zipkin/pull/105)
+- [OpenID Connect](/hub/kong-inc/openid-connect) (`openid-connect`)
+  - `kong-openid-connect` library updated to v2.1.0
+    - Updates `lua-resty-nettle` dependency to v2.0.
+    - Treats JWE tokens as opaque, so that they can be introspected.
```
Comment (eskibars) [inline app/enterprise/changelog.md:146]: This and the next line are really more like features and should be moved up to that section/out of bug fixes

File: app/enterprise/changelog.md [hunk 29]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+    - Adds support for Ed448 curve in EdDSA signing and verification and JWKS key generation.
+- [Kong JWT Signer](/hub/kong-inc/jwt-signer) (`jwt-signer`)
+  - Cache now uses upsert instead of insert/update with databases.
+  - Key rotation is now more resilient on errors.
```
Comment (eskibars) [inline app/enterprise/changelog.md:150]: More resilient how?

File: app/enterprise/changelog.md [hunk 30]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  - Adds DB-less improvements.
```
Comment (eskibars) [inline app/enterprise/changelog.md:151]: What kind of improvements?

**Review by eskibars — DISMISSED** [review]
Thanks for tagging me.  Left some comments/change requests.  Also cc @PaulFischer-Kong since many of these are fallout of the CE changelog

File: app/enterprise/changelog.md [hunk 31]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
 no_search: true
 no_version: true
 ---
+## 2.4.0.0 (beta)
+**Release Date** 2021/04/13
+
+### Features
+
+#### Core
+- This Kong Gateway version introduces relaxed version checks in hybrid mode between control planes
```
[the remark this answers is not in the document]
Comment (PaulFischer-Kong) [reply app/enterprise/changelog.md:12]: Under the headline - Managing the cluster using Control Plane nodes

https://docs.konghq.com/gateway-oss/2.4.x/hybrid-mode/?_ga=2.56159459.996419881.1618334217-1233986905.1614707525#managing-the-cluster-using-control-plane-nodes

File: app/enterprise/changelog.md [hunk 32]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  - `,` and `/` are reserved for filtering tags with "and" and "or", and are not allowed in tags.
+  - Non-printable ASCII (for example, the space character) is not allowed.
+- Kong Gateway now supports using an Online Certificate Status Protocol (OCSP) responder in the cluster for
+  hybrid mode control planes. This new feature can be configured in the `kong.conf` file.
+  [6887](https://github.com/Kong/kong/pull/6887)
+- Postgres `ssl_version` configuration now defaults to `any`. If `ssl_version` is not explicitly set,
+  the `any` option ensures that `luasec` will negotiate the most secure protocol available.
+
+#### PDK
+- This release includes a new JavaScript Plugin Development Kit (PDK). This addition
```
[the remark this answers is not in the document]
Comment (PaulFischer-Kong) [reply app/enterprise/changelog.md:32]: Javascript dev docs

https://docs.konghq.com/gateway-oss/2.4.x/external-plugins/?_ga=2.85453009.996419881.1618334217-1233986905.1614707525

File: app/enterprise/changelog.md (hunk 31 above)
[the remark this answers is not in the document]
Comment (HCloward) [reply app/enterprise/changelog.md:12]: I will add the link to the doc - Thanks @PaulFischer-Kong

File: app/enterprise/changelog.md (hunk 32 above)
[the remark this answers is not in the document]
Comment (HCloward) [reply app/enterprise/changelog.md:32]: I will add the link - Thanks @PaulFischer-Kong

File: app/enterprise/changelog.md [hunk 33]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  <div class="alert alert-ee blue">
+    Data planes are not allowed to connect to control planes if they are a different major version,
+    a version newer than the control plane’s version, or missing plugins from the control plane.
+  </div>
+- UTF-8 characters can now be used in tags. This update expands the range of
+  accepted characters in tags from a limited set of ASCII characters to almost all UTF-8 sequences.
+  Exceptions:
+  - `,` and `/` are reserved for filtering tags with "and" and "or", and are not allowed in tags.
+  - Non-printable ASCII (for example, the space character) is not allowed.
+- Kong Gateway now supports using an Online Certificate Status Protocol (OCSP) responder in the cluster for
```
[the remark this answers is not in the document]
Comment (HCloward) [reply app/enterprise/changelog.md:25]: Datong included some information in the in the [Configuration Reference](https://docs.konghq.com/gateway-oss/2.4.x/configuration/#cluster_ocsp). I will include a link.

File: app/enterprise/changelog.md [hunk 34]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  hybrid mode control planes. This new feature can be configured in the `kong.conf` file.
+  [6887](https://github.com/Kong/kong/pull/6887)
+- Postgres `ssl_version` configuration now defaults to `any`. If `ssl_version` is not explicitly set,
+  the `any` option ensures that `luasec` will negotiate the most secure protocol available.
+
+#### PDK
+- This release includes a new JavaScript Plugin Development Kit (PDK). This addition
+  allows users to write Kong plugins in JavaScript and TypeScript.
+- This release includes support for the Protobuf plugin communication protocol, which can be used in
```
[the remark this answers is not in the document]
Comment (HCloward) [reply app/enterprise/changelog.md:34]: I asked @javierguerragiraldez about this one in my Google doc I linked in the description for this PR and he said, "this is transparent to the users, even to developers of external plugins.   the only benefit is a possible performance improvement. contrary to google propaganda, protobuf isn't more portable or prevalent than other codecs.  the new JS and Python libraries use MessagePack.  the Kong core supports both equally." I am not sure if doc is needed here or not . . .

File: app/enterprise/changelog.md [hunk 35]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  place of MessagePack to communicate with non-Lua plugins. [6941](https://github.com/Kong/kong/pull/6941)
+- This release enables the `ssl_certificate` phase on plugins in the `stream` module.
+  [6873](https://github.com/Kong/kong/pull/6873)
+
+#### Plugins
+- **New plugin:** [OPA](/hub/kong-inc/opa) (`opa`)
+  - The OPA plugin forwards requests to an Open Policy Agent and processes the request
+    only if the authorization policy allows for it. **Released as BETA and should not be deployed
+    in production environment.**
+- **New plugin:** Mocking (`mocking`)
```
[the remark this answers is not in the document]
Comment (HCloward) [reply app/enterprise/changelog.md:44]: The docs are still being written for this as discussed in our meeting, so there is nothing to link it to as of yet.

File: app/enterprise/changelog.md (hunk 29 above)
Comment (HCloward) [reply app/enterprise/changelog.md:150]: This is the summary @bungle included in his PR - "Makes rotating keys more resilient on upstream endpoint errors and database errors and also make the code work better on dbless."

File: app/enterprise/changelog.md (hunk 30 above)
Comment (HCloward) [reply app/enterprise/changelog.md:151]: The only info I could find last night was what @bungle included in his PR commit messages which was, "Makes rotating keys more resilient on upstream endpoint errors and database errors and also make the code work better on dbless."

File: app/enterprise/changelog.md [hunk 36]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  still connect to the control plane. Also, data planes are allowed to have a superset
+  of plugins in addition to the control plane plugins. [6932](https://github.com/Kong/kong/pull/6932)
+  <div class="alert alert-ee blue">
+    Data planes are not allowed to connect to control planes if they are a different major version,
+    a version newer than the control plane’s version, or missing plugins from the control plane.
+  </div>
+- UTF-8 characters can now be used in tags. This update expands the range of
+  accepted characters in tags from a limited set of ASCII characters to almost all UTF-8 sequences.
+  Exceptions:
+  - `,` and `/` are reserved for filtering tags with "and" and "or", and are not allowed in tags.
```
[the remark this answers is not in the document]
Comment (HCloward) [reply app/enterprise/changelog.md:23]: There is some documentation on [Tags ](https://docs.konghq.com/gateway-oss/2.4.x/admin-api/#tags)in the Admin Api guide but it probably needs to be updated.

File: app/enterprise/changelog.md [hunk 37]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  - Kong administrators now have a powerful new capability to transform logs to any format that’s needed
+    for capturing, indexing, and correlating logs. When formatting, developers also have the ability to
+    execute custom Lua code, opening up a new range of possibilities for generating logs to whichever specification
+    administrators need the most. This feature uses the new PDK method `kong.log.set_serialize_value`,
+    as well as the new sandbox capability, both introduced in Kong v2.3. [6944](https://github.com/Kong/kong/pull/6944)
+
+### Fixes
+
+#### Core
+- Topological sort now prioritizes core, avoiding problems when plugin entities use core entities but
```
[the remark this answers is not in the document]
Comment (HCloward) [reply app/enterprise/changelog.md:71]: So this one is related to the "Migrations order is now guaranteed . . ." When I asked @kikito in the Google doc to explain the significance to the user he told me this, "Context: the only affected ones would be those who had plugins with migrations which depended on other plugins with migrations (so one had to run after the other). Usually our users don't setup their plugins like that - in part because, if they did, the random order in which migrations run before would give them trouble."

File: app/enterprise/changelog.md [hunk 38]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  entries. [6891](https://github.com/Kong/kong/pull/6891)
+- Migrations order is now guaranteed to be always the same. Before, the order in which the
+  subsystems were loaded changed randomly. Now, subsystems are sorted alphabetically, so the
+  order in which migrations are executed is constant. [6901](https://github.com/Kong/kong/pull/6901)
+- Buffered responses are disabled on connection upgrades. Kong cannot do buffered responses with,
+  for example, websocket connection upgrades. [6902](https://github.com/Kong/kong/pull/6902)
+- Entity relationships are now traverse-order-independent. This fix keeps track of all copied
+  records and injects the new field on each copy.
+  [6843](https://github.com/Kong/kong/commit/5f2a87259e3b474ec6129490bba82dac0aeba1cf)
+- The host header is updated between balancer retries. Before, Kong Gateway set the `Host` header during
```
[the remark this answers is not in the document]
Comment (HCloward) [reply app/enterprise/changelog.md:101]: ah, good catch, I will remove the other one then

File: app/enterprise/changelog.md [hunk 39]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+  to be retried. [6796](https://github.com/Kong/kong/pull/6796)
+- The router prioritizes the route with most matching headers when matching headers.
+  [6638](https://github.com/Kong/kong/pull/6638)
+- Fixed an issue involving multipart/form-data boundary checks. [6638](https://github.com/Kong/kong/pull/6638)
+- Fixed an issue that occurred when using upstreams for load balancing, where Kong was attempting
+  to resolve the upstream name instead of the hostname and failing with the errors `name resolution failed`
+  and `dns server error: 3 name error`. The following updates correct the issue:
+  - Kong does not cache empty upstream name dictionaries. [7002](https://github.com/Kong/kong/pull/7002)
+  - Kong does not assume upstreams don't exist after init phases. [7010](https://github.com/Kong/kong/pull/7010)
+- Opentracing libraries (opentracing, jaegar, datadog) bumped to latest versions.
```
[the remark this answers is not in the document]
Comment (HCloward) [reply app/enterprise/changelog.md:112]: yes, will do

File: app/enterprise/changelog.md (hunk 28 above)
Comment (HCloward) [reply app/enterprise/changelog.md:146]: Do you mean just the OpenID Connect updates or also the Kong JWT Signer updates? I will move them.

File: app/enterprise/changelog.md [hunk 40]
```diff
@@ -3,6 +3,156 @@ title: Kong Gateway (Enterprise) Changelog
...
+- Buffered responses are disabled on connection upgrades. Kong cannot do buffered responses with,
+  for example, websocket connection upgrades. [6902](https://github.com/Kong/kong/pull/6902)
+- Entity relationships are now traverse-order-independent. This fix keeps track of all copied
+  records and injects the new field on each copy.
+  [6843](https://github.com/Kong/kong/commit/5f2a87259e3b474ec6129490bba82dac0aeba1cf)
+- The host header is updated between balancer retries. Before, Kong Gateway set the `Host` header during
+  the access phase only and would send the wrong header if a target failed to serve and needed
+  to be retried. [6796](https://github.com/Kong/kong/pull/6796)
+- The router prioritizes the route with most matching headers when matching headers.
+  [6638](https://github.com/Kong/kong/pull/6638)
```
[the remark this answers is not in the document]
Comment (HCloward) [reply app/enterprise/changelog.md:105]: I will add the openresty update to the Dependency section I am creating - should I link to the changelog there?

**Review by GayleNeumann — APPROVED** [review]
LGTM. Shane had some questions that are not answered for this EE 2.4 beta changelog release, but we are tracking here: https://konghq.atlassian.net/browse/DOCU-1403 and will have these answered by EE 2.4 GA.
